src/services/analytics/TagManager.ts

```diff
@@ -0,0 +1,112 @@
+// Based on https://github.com/alinemorelli/react-gtm
+
+type DataLayer = Record<string, unknown>
+
+export type TagManagerArgs = {
+  /**
+   * GTM id, must be something like GTM-000000.
+   */
+  gtmId: string
+  /**
+   * Used to set environments.
+   */
+  auth?: string | undefined
+  /**
+   * Used to set environments, something like env-00.
+   */
+  preview?: string | undefined
+  /**
+   * Object that contains all of the information that you want to pass to Google Tag Manager.
+   */
+  dataLayer?: DataLayer | undefined
+}
+
+export const DATA_LAYER_NAME = 'dataLayer'
+
+export const _getRequiredGtmArgs = ({ gtmId, dataLayer = undefined, auth = '', preview = '' }: TagManagerArgs) => {
+  return {
+    gtmId,
+    dataLayer,
+    auth: auth ? `&gtm_auth=${auth}` : '',
+    preview: preview ? `&gtm_preview=${preview}` : '',
+  }
+}
+
+// Initialization scripts
+
+export const _getGtmScript = (args: TagManagerArgs) => {
+  const { gtmId, auth, preview } = _getRequiredGtmArgs(args)
+
+  const script = document.createElement('script')
+
+  const gtmScript = `
+    (function (w, d, s, l, i) {
+      w[l] = w[l] || [];
+      w[l].push({ 'gtm.start': new Date().getTime(), event: 'gtm.js' });
+      var f = d.getElementsByTagName(s)[0],
+        j = d.createElement(s),
+        dl = l != 'dataLayer' ? '&l=' + l : '';
+      j.async = true;
+      j.src = 'https://www.googletagmanager.com/gtm.js?id=' + i + dl + '${auth}${preview}&gtm_cookies_win=x';
+      f.parentNode.insertBefore(j, f);
+    })(window, document, 'script', '${DATA_LAYER_NAME}', '${gtmId}');`
+
+  script.innerHTML = gtmScript
+
+  return script
+}
+
+export const _getGtmNoScript = (args: TagManagerArgs) => {
+  const { gtmId, auth, preview } = _getRequiredGtmArgs(args)
+
+  const noscript = document.createElement('noscript')
+
+  const gtmIframe = `
+      <iframe src="https://www.googletagmanager.com/ns.html?id=${gtmId}${auth}${preview}&gtm_cookies_win=x"
+        height="0" width="0" style="display:none;visibility:hidden" id="tag-manager"></iframe>`
+
+  noscript.innerHTML = gtmIframe
+
+  return noscript
+}
```

Comment:
@iamacook our app doesn't work w/o JS, no need to include this.

Reply:
Fixed [here](https://github.com/safe-global/web-core/pull/538).